system/src/system_network.cpp

```diff
@@ -68,31 +69,156 @@ namespace {
 using namespace particle;
 using namespace spark;
 
-class RssiDiagnosticData: public AbstractIntegerDiagnosticData {
+static const system_tick_t SIGNAL_INFO_CACHE_INTERVAL = 1000;
+
+class SignalCache {
+public:
+    const Signal* getSignal() {
+#if Wiring_WiFi || Wiring_Cellular
+        system_tick_t m = millis();
+        if (ts_ == 0 || (m - ts_) >= SIGNAL_INFO_CACHE_INTERVAL) {
+# if Wiring_WiFi
+            sig_ = WiFi.RSSI();
+# elif Wiring_Cellular
+            sig_ = Cellular.RSSI();
+# endif
+            ts_ = millis();
+        }
+        return &sig_;
+#else
+        return nullptr;
+#endif
+    }
+private:
+#if Wiring_Cellular
+    CellularSignal sig_;
+#elif Wiring_WiFi
+    WiFiSignal sig_;
+#endif
+    system_tick_t ts_ = 0;
+};
+
```

Comment:
This is a good idea, since the higher level wiring APIs access the signal data field by field.